File: 0509-fibonacci-number/0509-fibonacci-number.js
/**
 * @param {number} n
 * @return {number}
 */
var fib = function(n) {
    if(n <= 1) return n;
    
    let cur = 0;
    let prev = 1;
    let prev2 = 0;
    
    for(let i = 2; i <= n; i++){
        cur = prev + prev2;
        
        prev2 = prev;
        prev = cur;
    }
    
    return cur;
};

File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    public int fib(int n) {
        int[] dp = new int[n+1];
        
        return helper(dp, n);
    }
    
    public int helper(int[] dp, int n){
        if (n == 0) return 0;
        if (n == 1) return 1;
        if(dp[n] != 0) return dp[n];
        
        dp[n] = helper(dp, n-1) + helper(dp, n-2);
        return dp[n];
    }
}